File: rust/parity/specs/vyper_external_call_sentry.py
"""Parity spec for PiVyperExternalCallSentry.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiVyperExternalCallSentry"

_mod = load_py_agent("pi_vyper_external_call_sentry.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiVyperExternalCallSentry()
    out = agent.audit_vyper_external_call(_mod.VyperExternalCallInput(**data))
    return out.model_dump()


# Clean: external call AFTER state change (Checks-Effects-Interactions respected)
_SECURE = "\n".join(
    [
        "@external",
        "def withdraw(amount: uint256):",
        "    self.balance -= amount",
        "    raw_call(msg.sender, b'')",
    ]
)

# Vulnerable via raw_call then `self.x = ...`
_REENTRANT_ASSIGN = "\n".join(
    [
        "@external",
        "def withdraw(amount: uint256):",
        "    raw_call(msg.sender, b'')",
        "    self.balance = 0",
    ]
)

# Vulnerable via ext_call then augmented assignment `+=`
_REENTRANT_AUGMENTED = "\n".join(
    [
        "@nonreentrant('lock')",
        "@external",
        "def claim():",
        "    ext_call(self.target)",
        "    self.total += 1",
    ]
)

# Comment lines must be ignored; here the only state mod is in a comment so SECURE
_COMMENTS_IGNORED = "\n".join(
    [
        "@external",
        "def safe():",
        "    raw_call(msg.sender, b'')",
        "    # self.balance = 0  (this is just a comment)",
        "    return True",
    ]
)

# Multiple functions, one secure + one vulnerable
_MIXED = "\n".join(
    [
        "@view",
        "def get_balance() -> uint256:",
        "    return self.balance",
        "",
        "@external",
        "def drain():",
        "    raw_call(attacker, b'')",
        "    self.balance = 0",
    ]
)

# No external call at all -> secure regardless of state writes
_NO_EXT_CALL = "\n".join(
    [
        "@external",
        "def setup():",
        "    self.owner = msg.sender",
        "    self.total = 100",
    ]
)

# No function definitions at all
_NO_FUNCS = "# just a module-level comment\nWAD: constant(uint256) = 10 ** 18"

SAMPLES = [
    {"input": {"file_path": "vault.vy", "vyper_code": _SECURE}},
    {"input": {"file_path": "vault.vy", "vyper_code": _REENTRANT_ASSIGN}},
    {"input": {"file_path": "vault.vy", "vyper_code": _REENTRANT_AUGMENTED}},
    {"input": {"file_path": "vault.vy", "vyper_code": _COMMENTS_IGNORED}},
    {"input": {"file_path": "vault.vy", "vyper_code": _MIXED}},
    {"input": {"file_path": "vault.vy", "vyper_code": _NO_EXT_CALL}},
    {"input": {"file_path": "vault.vy", "vyper_code": _NO_FUNCS}},
    {"input": {"file_path": "vault.vy", "vyper_code": ""}},
    {"input": {"file_path": "vault.vy", "vyper_code": _REENTRANT_ASSIGN, "check_level": "MEDIUM"}},
    # non-strict env -> WARN path, is_secure coerced back to True
    {"input": {"file_path": "vault.vy", "vyper_code": _REENTRANT_ASSIGN},
     "env": {"PI_VYPER_EXTERNAL_CALL_STRICT_MODE": "false"}},
    # explicit strict env -> REJECTED path
    {"input": {"file_path": "vault.vy", "vyper_code": _REENTRANT_AUGMENTED},
     "env": {"PI_VYPER_EXTERNAL_CALL_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/vyper_external_call_sentry.rs
//! Port of `pi_micro_agents/pi_vyper_external_call_sentry.py`.
//!
//! Audits Vyper source code to ensure external calls occur after state changes
//! (Checks-Effects-Interactions). Behaviour is a line-for-line mirror of the
//! Python original.

use crate::pyutil;
use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub file_path: String,
    pub vyper_code: String,
    #[serde(default = "default_check_level")]
    pub check_level: String,
}

fn default_check_level() -> String {
    "STRICT".to_string()
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub vulnerable_functions: Vec<String>,
    pub flagged_findings: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true".
///
/// NOTE: the Python original additionally falls back to reading
/// `~/.antigravitycli/config.json` (and a sibling repo path) when the env var is
/// unset, defaulting to `True` on any error. The parity harness always exercises
/// this agent with the env var set, and on a clean checkout no config file is
/// present, so the env-var branch plus the `True` default fully reproduces the
/// observed behaviour. The config-file branch is NOT replicated here (filesystem
/// I/O is intentionally excluded from the deterministic port).
fn is_strict_mode() -> bool {
    match std::env::var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

// Header portion of the Python function-block regex (everything up to and
// including the `\s*` that precedes the body group). The body group and its
// lookahead are handled manually in `find_func_blocks`, because the `regex`
// crate has no lookahead support.
//
// Python: r'((?:@[a-zA-Z0-9_]+(?:\([^)]*\))?\s*)*)def\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^:]*:\s*([\s\S]*?)(?=\n\S|\Z)'
// The greedy `\s*` after the colon is kept in this header so its
// (parity-critical) whitespace consumption matches the original.
static HEADER_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"((?:@[a-zA-Z0-9_]+(?:\([^)]*\))?\s*)*)def\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^:]*:\s*")
        .unwrap()
});

// State-modification patterns, ported verbatim from the Python list.
static STATE_MOD_RES: Lazy<Vec<Regex>> = Lazy::new(|| {
    vec![
        Regex::new(r"self\.[a-zA-Z0-9_]+\s*=").unwrap(),
        Regex::new(r"self\.[a-zA-Z0-9_]+\s*(\+=|-=)").unwrap(),
    ]
});

/// True if `c` is whitespace per the regex `\s` class (which Python's `\S`
/// negates). Rust's `char::is_whitespace` omits `\x1c\x1d\x1e`, which Python's
/// `re` `\s` (on `str`) treats as whitespace, so handle them explicitly.
fn is_regex_whitespace(c: char) -> bool {
    c.is_whitespace() || matches!(c, '\u{1c}' | '\u{1d}' | '\u{1e}')
}

/// Compute the end byte offset of a function body that starts at `start`,
/// reproducing the Python lookahead `(?=\n\S|\Z)`: stop at the first `\n`
/// immediately followed by a non-whitespace character, else end-of-string.
fn body_end(code: &str, start: usize) -> usize {
    let bytes = code.as_bytes();
    let n = code.len();
    let mut i = start;
    while i < n {
        if bytes[i] == b'\n' {
            if i + 1 < n {
                let next = code[i + 1..].chars().next().unwrap();
                if !is_regex_whitespace(next) {
                    return i;
                }
            }
        }
        i += 1;
    }
    n
}

/// Reproduces `re.findall(...)` for the function-block pattern, returning
/// `(decorators, name, args, body)` tuples in source order, non-overlapping.
fn find_func_blocks(code: &str) -> Vec<(String, String, String, String)> {
    let mut out: Vec<(String, String, String, String)> = Vec::new();
    let mut pos = 0usize;
    let n = code.len();
    while pos <= n {
        let m = match HEADER_RE.find_at(code, pos) {
            Some(m) => m,
            None => break,
        };
        let caps = HEADER_RE.captures_at(code, pos).unwrap();
        let decorators = caps.get(1).map_or("", |c| c.as_str()).to_string();
        let name = caps.get(2).map_or("", |c| c.as_str()).to_string();
        let args = caps.get(3).map_or("", |c| c.as_str()).to_string();
        let body_start = m.end();
        let body_stop = body_end(code, body_start);
        let body = code[body_start..body_stop].to_string();
        out.push((decorators, name, args, body));

        // Advance like `re.findall`: continue from the end of the full match,
        // which is the end of the body (the lookahead is zero-width). Guard
        // against non-advancing matches (mirrors re's empty-match bump).
        let mut newpos = if body_stop > m.end() { body_stop } else { m.end() };
        if newpos <= m.start() {
            newpos = m.start() + 1;
        }
        pos = if newpos > pos { newpos } else { pos + 1 };
    }
    out
}

pub fn audit_vyper_external_call(input: &Input) -> Output {
    let code = &input.vyper_code;
    let mut vulnerable_funcs: Vec<String> = Vec::new();
    let mut flagged_findings: Vec<String> = Vec::new();

    let func_blocks = find_func_blocks(code);

    for (_decorators, name, _args, body) in &func_blocks {
        let lines = pyutil::splitlines(body);
        let mut external_call_seen = false;
        let mut first_ext_call_line: i64 = -1;

        for (idx, line) in lines.iter().enumerate() {
            let line_stripped = pyutil::strip(line);
            // Exclude comments
            if line_stripped.starts_with('#') {
                continue;
            }

            // Check if this line is an external call
            if line_stripped.contains("ext_call") || line_stripped.contains("raw_call") {
                external_call_seen = true;
                if first_ext_call_line == -1 {
                    first_ext_call_line = idx as i64;
                }
            }

            // Check if state change happens after external call is seen
            if external_call_seen {
                let hit = STATE_MOD_RES.iter().any(|re| re.is_match(line_stripped));
                if hit {
                    vulnerable_funcs.push(name.clone());
                    flagged_findings.push(format!(
                        "Function '{name}' modifies local state in line {} ('{line_stripped}') \
after an external call was executed in line {}. \
This violates the Checks-Effects-Interactions pattern and introduces a potential reentrancy vulnerability.",
                        idx + 1,
                        first_ext_call_line + 1
                    ));
                    break;
                }
            }
        }
    }

    let mut is_secure = vulnerable_funcs.is_empty();
    let risk_score = if !is_secure { 80.0 } else { 0.0 };

    let is_strict = is_strict_mode();
    let mut status = "PASSED".to_string();
    if !is_secure {
        if is_strict {
            status = "REJECTED_VYPER_CALL_RISK".to_string();
        } else {
            status = "WARN_VYPER_CALL_RISK".to_string();
            is_secure = true;
        }
    }

    Output {
        is_secure,
        vulnerable_functions: vulnerable_funcs,
        flagged_findings,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = audit_vyper_external_call(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serial_test::serial;

    fn run(code: &str) -> Output {
        audit_vyper_external_call(&Input {
            file_path: "f.vy".into(),
            vyper_code: code.into(),
            check_level: "STRICT".into(),
        })
    }

    #[test]
    #[serial]
    fn secure_code_passes() {
        std::env::set_var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE", "true");
        let o = run("@view\ndef get_balance() -> uint256:\n    return self.balance\n");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.vulnerable_functions.is_empty());
        std::env::remove_var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE");
    }

    #[test]
    #[serial]
    fn reentrancy_flagged_strict() {
        std::env::set_var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE", "true");
        let o = run("@external\ndef withdraw(amount: uint256):\n    raw_call(msg.sender, b\"\")\n    self.balance = 0\n");
        assert!(!o.is_secure);
        assert_eq!(o.status, "REJECTED_VYPER_CALL_RISK");
        assert_eq!(o.risk_score, 80.0);
        assert_eq!(o.vulnerable_functions, vec!["withdraw"]);
        std::env::remove_var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE");
    }

    #[test]
    #[serial]
    fn reentrancy_warn_non_strict() {
        std::env::set_var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE", "false");
        let o = run("@external\ndef withdraw():\n    ext_call()\n    self.x += 1");
        // non-strict coerces is_secure back to True
        assert!(o.is_secure);
        assert_eq!(o.status, "WARN_VYPER_CALL_RISK");
        assert_eq!(o.risk_score, 80.0);
        assert_eq!(o.vulnerable_functions, vec!["withdraw"]);
        std::env::remove_var("PI_VYPER_EXTERNAL_CALL_STRICT_MODE");
    }
}
